[macOS] -webkit-user-select: none disables searching in Safari.

megangardner opened the pull request:
#### 369438f43a82604e7e9209225d10d7abfb9a5219
<pre>
[macOS] -webkit-user-select: none disables searching in Safari.
<a href="https://bugs.webkit.org/show_bug.cgi?id=298345">https://bugs.webkit.org/show_bug.cgi?id=298345</a>
<a href="https://rdar.apple.com/8081660">rdar://8081660</a>

Reviewed by Aditya Keerthi.

Follow up work for mac in relation to 298785@main.

We do not have the fallback range available in the same
way on macOS as we did on iOS. In order to have a similar
solution for creating an indicator based on the range,
we need to pass that range back to the FindController so
that it can be used in the fallback case.

We also need to be checking the normalized range,
as the regular range will return the uncollapsed version,
and that will not send us down the fallback path.

* Source/WebCore/editing/Editor.cpp:
(WebCore::Editor::findString):
* Source/WebCore/editing/Editor.h:
* Source/WebCore/editing/EditorCommand.cpp:
(WebCore::executeFindString):
* Source/WebCore/page/LocalDOMWindow.cpp:
(WebCore::LocalDOMWindow::find const):
* Source/WebCore/page/Page.cpp:
(WebCore::Page::findString):
* Source/WebCore/page/Page.h:
* Source/WebKit/WebProcess/WebPage/FindController.cpp:
(WebKit::FindController::frameWithSelection):
(WebKit::FindController::findString):
(WebKit::FindController::updateFindIndicator):
* Source/WebKit/WebProcess/WebPage/FindController.h:

Canonical link: <a href="https://commits.webkit.org/299624@main">https://commits.webkit.org/299624@main</a>
</pre>
<!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/54ef7cd122de2d993114727e33938b25383444ff

| Misc | iOS, visionOS, tvOS & watchOS  | macOS  | Linux |  Windows | Apple Internal |
| ----- | ---------------------- | ------- |  ----- |  --------- | ------ |
| [✅ 🧪 style](https://ews-build.webkit.org/#/builders/38/builds/119549 "Passed style check") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/131/builds/39241 "Built successfully") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/138/builds/29895 "Built successfully") | [✅ 🛠 wpe](https://ews-build.webkit.org/#/builders/5/builds/125822 "Built successfully") | [✅ 🛠 win](https://ews-build.webkit.org/#/builders/59/builds/71623 "Built successfully") | [✅ 🛠 ios-apple](https://ews-bridge.webkit.apple.com/builds/sw/T-276/92699e7b-e2c4-41dd-8b79-0fc7c00d5b52/b2f02738-1cd2-43b5-b754-a625def3b88b) 
| [✅ 🧪 bindings](https://ews-build.webkit.org/#/builders/9/builds/121425 "Passed tests") | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/130/builds/39939 "Built successfully") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/123/builds/47821 "Built successfully") | [✅ 🧪 wpe-wk2](https://ews-build.webkit.org/#/builders/34/builds/90810 "Passed tests") | [✅ 🧪 win-tests](https://ews-build.webkit.org/#/builders/60/builds/60115 "Passed tests") | [✅ 🛠 mac-apple](https://ews-bridge.webkit.apple.com/builds/sw/T-276/92699e7b-e2c4-41dd-8b79-0fc7c00d5b52/94cb8f04-2db0-4c29-8e4b-c71d8112b7ee) 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/11/builds/122500 "Passed tests") | [✅ 🧪 ios-wk2](https://ews-build.webkit.org/#/builders/132/builds/31878 "Passed tests") | [✅ 🧪 api-mac](https://ews-build.webkit.org/#/builders/18/builds/107196 "Passed tests") | [❌ 🧪 api-wpe](https://ews-build.webkit.org/#/builders/41/builds/71304 "Found 1 new API test failure: WPE/TestWebKitWebContext:/webkit/WebKitWebContext/no-web-process-leak (failure)") | | [✅ 🛠 vision-apple](https://ews-bridge.webkit.apple.com/builds/sw/T-276/92699e7b-e2c4-41dd-8b79-0fc7c00d5b52/4b495079-2ca9-4478-ade9-c5a19ce39032) 
| | [✅ 🧪 ios-wk2-wpt](https://ews-build.webkit.org/#/builders/133/builds/30915 "Passed tests") | [✅ 🧪 mac-wk1](https://ews-build.webkit.org/#/builders/135/builds/25301 "Passed tests") | [✅ 🛠 wpe-cairo](https://ews-build.webkit.org/#/builders/65/builds/69471 "Built successfully") | | 
| | [✅ 🧪 api-ios](https://ews-build.webkit.org/#/builders/13/builds/101342 "Passed tests") | [✅ 🧪 mac-wk2](https://ews-build.webkit.org/#/builders/136/builds/25494 "Passed tests") | [✅ 🛠 gtk](https://ews-build.webkit.org/#/builders/2/builds/128797 "Built successfully") | | 
| | [✅ 🛠 vision](https://ews-build.webkit.org/#/builders/128/builds/46471 "Built successfully") | [✅ 🧪 mac-AS-debug-wk2](https://ews-build.webkit.org/#/builders/122/builds/35194 "Passed tests") | [✅ 🧪 gtk-wk2](https://ews-build.webkit.org/#/builders/1/builds/99402 "Passed tests") | | 
| | [✅ 🛠 vision-sim](https://ews-build.webkit.org/#/builders/121/builds/46836 "Built successfully") | [✅ 🧪 mac-wk2-stress](https://ews-build.webkit.org/#/builders/8/builds/103390 "Passed tests") | [✅ 🧪 api-gtk](https://ews-build.webkit.org/#/builders/21/builds/99233 "Passed tests") | | 
| [✅ 🛠 🧪 merge](https://ews-build.webkit.org/#/builders/19/builds/25224 "Built successfully and passed tests") | [✅ 🧪 vision-wk2](https://ews-build.webkit.org/#/builders/126/builds/44669 "Passed tests") | [✅ 🧪 mac-intel-wk2](https://ews-build.webkit.org/#/builders/137/builds/22684 "Passed tests") | [✅ 🛠 playstation](https://ews-build.webkit.org/#/builders/134/builds/43035 "Built successfully") | | 
| | [✅ 🛠 tv](https://ews-build.webkit.org/#/builders/127/builds/46333 "Built successfully") | [✅ 🛠 mac-safer-cpp](https://ews-build.webkit.org/#/builders/120/builds/52039 "Built successfully") | | | 
| | [✅ 🛠 tv-sim](https://ews-build.webkit.org/#/builders/125/builds/45798 "Built successfully") | | | | 
| | [✅ 🛠 watch](https://ews-build.webkit.org/#/builders/129/builds/49148 "Built successfully") | | | | 
| | [✅ 🛠 watch-sim](https://ews-build.webkit.org/#/builders/124/builds/47485 "Built successfully") | | | | 
<!--EWS-Status-Bubble-End-->

pxlcoder commented on Source/WebCore/page/Page.h:
```diff
@@ -555,7 +555,11 @@ class Page : public RefCountedAndCanMakeWeakPtr<Page>, public Supplementable<Pag
+        std::optional<FrameIdentifier> frameIdentifer;
+        std::optional<SimpleRange> range;
+    };
+    WEBCORE_EXPORT FindStringData findString(const String&, FindOptions, DidWrap* = nullptr);
```
It seems like we could make `findString` return an `optional<FindStringData>`, and make the members of `FindStringData` non-optional?

megangardner replied:
But one if you get one and not the other? It seems like we shouldn't block on that, even if we don't see a reason for it now.

pxlcoder replied:
Is that possible?

megangardner replied:
Probably not, but if we leave it in this form, we can remove the out-parameter as pointed out by Richard below, which I think is also valuable.

rr-codes replied:
(I agree more with Aditya than my own suggestion fwiw)

pxlcoder commented on Source/WebCore/page/Page.cpp:
```diff
@@ -1195,13 +1195,14 @@ std::optional<FrameIdentifier> Page::findString(const String& target, FindOption
         }
-        if (localFrame->protectedEditor()->findString(target, (options - FindOption::WrapAround) | FindOption::StartInSelection)) {
+        auto foundRange = localFrame->protectedEditor()->findString(target, (options - FindOption::WrapAround) | FindOption::StartInSelection);
+        if (foundRange.has_value()) {
```
```suggestion
        if (foundRange) {
```

rr-codes commented on Source/WebCore/page/Page.h:
```diff
@@ -555,7 +555,11 @@ class Page : public RefCountedAndCanMakeWeakPtr<Page>, public Supplementable<Pag
 
-    WEBCORE_EXPORT std::optional<FrameIdentifier> findString(const String&, FindOptions, DidWrap* = nullptr);
+    struct FindStringData {
+        std::optional<FrameIdentifier> frameIdentifer;
```
```suggestion
        std::optional<FrameIdentifier> frameIdentifier;
```

rr-codes commented on Source/WebCore/page/Page.h:
```diff
@@ -555,7 +555,11 @@ class Page : public RefCountedAndCanMakeWeakPtr<Page>, public Supplementable<Pag
+        std::optional<FrameIdentifier> frameIdentifer;
+        std::optional<SimpleRange> range;
+    };
+    WEBCORE_EXPORT FindStringData findString(const String&, FindOptions, DidWrap* = nullptr);
```
since the result is a struct now anyways, the DidWrap out parameter can just be part of the result directly instead

pxlcoder replied:
Unless we do this: https://github.com/WebKit/WebKit/pull/50265#discussion_r2320699935

rr-codes replied:
oops ya, missed that comment

megangardner replied:
I hate an out-param and much as the next person, but this is a thing that is done on many levels of this code, I or someone else should fix this as a followup.

pxlcoder commented on Source/WebKit/WebProcess/WebPage/FindController.h:
```diff
@@ -128,6 +129,10 @@ class FindController final : private WebCore::PageOverlayClient {
     int m_foundStringMatchIndex { -1 };
 
+    // For instances where the normalized selection range is collapsed.
+    std::optional<WebCore::SimpleRange> m_lastFoundRange;
+    std::optional<WebCore::VisibleSelection> m_lastSelection;
```
Can we reset these to `nullopt` in `FindController::hideFindUI`?

megangardner committed:
[macOS] -webkit-user-select: none disables searching in Safari.
https://bugs.webkit.org/show_bug.cgi?id=298345
rdar://8081660

Reviewed by Aditya Keerthi.

Follow up work for mac in relation to 298785@main.

We do not have the fallback range available in the same
way on macOS as we did on iOS. In order to have a similar
solution for creating an indicator based on the range,
we need to pass that range back to the FindController so
that it can be used in the fallback case.

We also need to be checking the normalized range,
as the regular range will return the uncollapsed version,
and that will not send us down the fallback path.

* Source/WebCore/editing/Editor.cpp:
(WebCore::Editor::findString):
* Source/WebCore/editing/Editor.h:
* Source/WebCore/editing/EditorCommand.cpp:
(WebCore::executeFindString):
* Source/WebCore/page/LocalDOMWindow.cpp:
(WebCore::LocalDOMWindow::find const):
* Source/WebCore/page/Page.cpp:
(WebCore::Page::findString):
* Source/WebCore/page/Page.h:
* Source/WebKit/WebProcess/WebPage/FindController.cpp:
(WebKit::FindController::frameWithSelection):
(WebKit::FindController::findString):
(WebKit::FindController::updateFindIndicator):
* Source/WebKit/WebProcess/WebPage/FindController.h:

Canonical link: https://commits.webkit.org/299624@main
